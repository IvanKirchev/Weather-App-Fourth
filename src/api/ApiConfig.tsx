export default {
    openweatherUrl: 'https://api.openweathermap.org',
    // TODO don't store secrets in source control
    openWeatherAppID: '3045bd446337c6e9aa7ef0719614442d'
}