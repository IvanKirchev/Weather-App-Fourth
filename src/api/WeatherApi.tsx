import axios from "axios";
import ApiConfig from "./ApiConfig";

interface IForecast {
    city: ICity
    days: IDayForecast[]
    list: any
}

interface IDayForecast {
    id: number
    name: string,
    averageTemp: number
    forecastByHours: IHourForecast[]
    icon: string,
    minTemp: number,
    maxTemp: number,
    weatherDescription: string
}

interface IHourForecast {
    dt: number
    main: IMainForecastData
    weather: IWeatherDescription[]
    wind: IWind
    visibility: number
}

interface IMainForecastData {
        temp: number
        feels_like: number
        temp_min: number
        temp_max: number
        pressure: number
        sea_level: number
        grnd_level: number
        humidity: number
        temp_kf: number
}

interface IWeatherDescription {
    id: number
    main: string
    description: string
    icon: string
}

interface IWind {
    speed: number
    deg: number
    gust: number
}

interface ICity {
    country: string
    id: number
    name: string
    population: number
    sunrise: number
    sunset: number
    timezone: number
}

function getForecastByCity(city: string, callback: (data: IForecast | null, error: unknown | null) => void) {
    const fetchData = async (): Promise<void> => {
        try {
            const response = await axios.get(`${ApiConfig.openweatherUrl}/data/2.5/forecast?q=${city}&appid=${ApiConfig.openWeatherAppID}&units=metric`)
            callback(response.data, null)
        } catch(error: unknown) {
            callback(null, error)
        }
    };

    fetchData();
}

export { getForecastByCity }
export type { IForecast, IDayForecast, IHourForecast, IMainForecastData, IWeatherDescription, IWind, ICity };
