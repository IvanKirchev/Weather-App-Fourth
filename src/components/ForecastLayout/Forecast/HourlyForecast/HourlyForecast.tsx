import { Box, Grid, Paper, styled, Typography } from "@mui/material";
import react from "react";
import { IHourForecast } from "../../../../api/WeatherApi";

interface IHourForecastProps {
  hourForecast: IHourForecast | null;
}

function HourlyForecast(props: IHourForecastProps) {
  const Img = styled("img")({
    margin: "auto",
    display: "block",
    width: "80%",
    position: "relative",
    top: "12%",
  });

  return (
    <div>
      <Box textAlign="left" sx={{ width: "55rem", marginLeft: "27%" }}>
        <Paper sx={{ p: 2, margin: "auto", maxWidth: 500, flexGrow: 1 }}>
          <Grid container spacing={2}>
            <Grid item>
              <Typography sx={{ width: 128, height: 128 }}>
                <Img
                  alt="complex"
                  src={`http://openweathermap.org/img/w/${props.hourForecast?.weather[0].icon}.png`}
                />
              </Typography>
            </Grid>
            <Grid item xs={12} sm container>
              <Grid item xs container direction="column" spacing={2}>
                <Grid item xs>
                  <Typography gutterBottom variant="h5" component="div">
                    {props.hourForecast
                      ? Math.round(props.hourForecast?.main.temp * 10) / 10
                      : 0}{" "}
                    °C
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    Feels like{" "}
                    {props.hourForecast
                      ? Math.round(props.hourForecast?.main.feels_like * 10) /
                        10
                      : 0}{" "}
                    °C
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    Humidity: {props.hourForecast?.main.humidity}
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    Wind: {props.hourForecast?.wind.speed}
                  </Typography>
                  <Typography variant="body2" color="text.secondary">
                    Pressure: {props.hourForecast?.main.pressure}
                  </Typography>
                </Grid>
              </Grid>
              <Grid item xs>
                <Typography variant="h6" component="div">
                  {props.hourForecast?.weather[0].description}
                </Typography>
                <Typography
                  variant="caption"
                  color="text.secondary"
                  position="relative"
                  top="56%"
                  left="35%"
                >
                  {props.hourForecast
                    ? new Date(
                        props.hourForecast?.dt * 1000
                      ).toLocaleDateString([], {
                        hour: "2-digit",
                        minute: "2-digit",
                        hour12: false,
                        timeZoneName: "short"
                      })
                    : null}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
        </Paper>
      </Box>
    </div>
  );
}

export default HourlyForecast;
