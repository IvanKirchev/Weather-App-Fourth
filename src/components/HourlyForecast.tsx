import { Box, Slider } from "@mui/material";
import react, { Dispatch, SetStateAction, useEffect } from "react"
import { IDayForecast, IHourForecast } from "../api/WeatherApi";
import DayCard from "./DayCard";

interface IHourlyForecastProps {
    day: IDayForecast | null
    selectedHour: IHourForecast | null
    setSelectedHour: Dispatch<SetStateAction<IHourForecast | null>>
    selectedHourIndex: number,
    setSelectedHourIndex: Dispatch<SetStateAction<number>>
}

function HourlyForecast(props: IHourlyForecastProps) {

    function getHourByIndex(value: number) {
        if(props.day) {
            let date = new Date(props.day?.forecastByHours[value].dt * 1000)
            return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit', timeZoneName: "short" });
        } else {
            return ""
        }
    }

    const handleChange = (event: Event, newValue: number | number[]) => {
        if (typeof newValue === 'number') {
            if(props.day?.forecastByHours[newValue] != undefined) {
                let selected = props.day?.forecastByHours[newValue];
                props.setSelectedHour(selected);
                // When the slider is changed (value change or rerender) set the current value to the last avaible index if the new value is not available
                //console.log(newValue)
                props.setSelectedHourIndex(props.day.forecastByHours.length > newValue ? newValue : props.day.forecastByHours.length -1);
            }
        }
      };

      function getSelectedHourIndex(): number {
          if(props.selectedHour) {
            let index = props.day?.forecastByHours.indexOf(props.selectedHour);
            return index ? index : 0;
          }
          return 0
      }

    return (
        <>
        {props.day ? (
            <div>
                <Box sx={{ width: "41rem", marginLeft: "34%"}}>
                    <Slider
                        aria-label="Temperature"
                        value={props.selectedHourIndex}
                        
                        valueLabelDisplay="off"
                        step={1}
                        marks={props.day.forecastByHours.map(h => {
                            let index = props.day?.forecastByHours.indexOf(h);
                            index = index ? index : 0;
                            return {
                                value: index,
                                label: getHourByIndex(index)
                            }
                        })}
                        min={0}
                        max={props.day.forecastByHours.length - 1 === 0 ? 1 : props.day.forecastByHours.length - 1}
                        onChange={handleChange}
                    />
                </Box>
            </div>
        ) : null }
        
        </>
    );
}

export default HourlyForecast;