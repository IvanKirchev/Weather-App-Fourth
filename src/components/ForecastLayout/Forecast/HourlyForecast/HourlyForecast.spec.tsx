import React from "react";
import renderer from "react-test-renderer";
import HourlyForecast from "./HourlyForecast";
import enzyme from "enzyme";
import { IHourForecast } from "../../../../api/WeatherApi";


const mockHourForecast: IHourForecast = {
    dt:16372368,
    main: {
        temp: 13,
        feels_like: 13,
        temp_min: 13,
        temp_max: 13,
        pressure: 13,
        sea_level: 13,
        grnd_level: 13,
        humidity: 13,
        temp_kf: 13
    },
    weather: [
        {
            id: 13,
            main: "13",
            description: "Nice weather",
            icon: "icon url",
        }
    ],
    wind: {
        speed: 13,
        deg: 13,
        gust: 14
    },
    visibility: 1000
}

it("Should render HourlyForecast component correctly", () => {
    const component = renderer.create(<HourlyForecast hourForecast={mockHourForecast}/>)
    expect(component).toMatchSnapshot();
})