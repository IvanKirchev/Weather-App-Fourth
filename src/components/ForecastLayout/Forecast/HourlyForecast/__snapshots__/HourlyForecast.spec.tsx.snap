// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render HourlyForecast component correctly 1`] = `
<div>
  <div
    className="MuiBox-root css-15jbnc6"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1s8qwjp-MuiPaper-root"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-152sepb-MuiTypography-root"
          >
            <img
              alt="complex"
              className="css-zbzyxd"
              src="http://openweathermap.org/img/w/icon url.png"
            />
          </p>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-undefined MuiGrid-grid-xs-12 MuiGrid-grid-sm-true css-19gf1bx-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-grid-xs-true css-1x3r9m2-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
            >
              <div
                className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                13
                 
                °C
              </div>
              <p
                className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                Feels like
                 
                13
                 
                °C
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                Humidity: 
                13
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                Wind: 
                13
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                Pressure: 
                13
              </p>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Nice weather
            </div>
            <span
              className="MuiTypography-root MuiTypography-caption css-1jn4ul4-MuiTypography-root"
            >
              7/9/1970, 13:52 GMT+2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
